### PR TITLE
OIDC & hosted-lock support

### PR DESCRIPTION
**Pull request by markrcote**
This swaps out the old Auth0 login stuff for the new hosted-lock OIDC solution.

I also did a bit of cleanup and upgraded jquery, bootstrap, and datatables, which got rid of the last few CSP errors, so I could tighten those polices.

### CODE REVIEW

**Comment by edmorley** [comment]
(Currently the PR is a massive +650 −7,738 hehe)

**Comment by edmorley** [comment]
Hmm I also see some low hanging fruit for the Travis run that would be good for someone to address in the future, eg:
* switching from precise to trusty (using `dist: trusty`)
* adding flake8

**Comment by markrcote** [comment]
Sure thing! I split out about half of the commits into a separate branch. I left the CSRF protection on logout in this branch; hope that's all right. :)

Only +170 / -228 now! :)

**Comment by markrcote** [comment]
Also I encourage you to review the commits one-by-one. I tried to keep the changes in each logically separate and for them to tell a story as they go. :)

File: pulseguardian/web.py [hunk 1]
```diff
@@ -491,99 +484,9 @@ def bindings_listing(queue_name):
...
-    )
-    token_payload = {
-        'client_id': config.auth0_client_id,
-        'client_secret': config.auth0_client_secret,
-        'redirect_uri': config.auth0_callback_url,
-        'code': code,
-        'grant_type': 'authorization_code',
-    }
-    token_info = requests.post(token_url,
-                               data=json.dumps(token_payload),
```
Comment (edmorley) [inline pulseguardian/web.py]: ./pulseguardian/web.py:34:1: F401 'json' imported but unused

(Also fwiw, the `json.dumps()` here was unnecessary - requests has a `json` argument that does this automatically)

File: README.md [hunk 2]
```diff
@@ -101,15 +100,17 @@ reviewer can quickly verify that all tests still pass with your changes.
...
 * Run the Pulse Guardian daemon with: `python pulseguardian/guardian.py`
 * Run the web app (for development) with: `python pulseguardian/web.py`
 * For production, the web app can be run with [gunicorn][] and such.
 
-The `FAKE_ACCOUNT` variable will make development easier. This feature will
-disable HTTPS and bypass Auth0 for testing. It will also create the
-given user, if necessary, and log in automatically.
+Setting `FAKE_ACCOUNT` to an email address will make development easier. This
+feature bypasses OIDC authentication, logging the user in with the provided
+address .  It will also create the given user, if necessary.
```
Comment (edmorley) [inline README.md]: `s/address /address/`

**Comment by markrcote** [comment]
Thanks. :) Sorry for the couple of unused imports; I switched computers and just remembered that I haven't set up all my linters. Also further reason for me to set up flake8 on Travis, as you suggested. :)

File: requirements.txt [hunk 3]
```diff
@@ -16,3 +16,4 @@ Flask-SSLify==0.1.5
 alembic==0.8.3
 setuptools==36.0.1
 flask-secure-headers==0.6
+Flask-pyoidc==1.1.0
```
Comment (markrcote) [inline requirements.txt]: I was trying to figure out if it is good practice to pin every requirement's dependencies. Seems there is some disagreement. As for PulseGuardian, I think some are pinned and some aren't, so the worst of both worlds, heh.

File: pulseguardian/auth.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            authorization_endpoint=self.oidc_config.auth_endpoint(),
+            token_endpoint=self.oidc_config.token_endpoint(),
+            userinfo_endpoint=self.oidc_config.userinfo_endpoint(),
+        )
+
+    def auth(self, app):
+        if config.fake_account:
+            return FakeOIDCAuthentication()
+
+        oidc = OIDCAuthentication(
```
Comment (markrcote) [inline pulseguardian/auth.py]: Awesome, yeah that seems to work fine!

File: requirements.txt (hunk 3 above)
Comment (markrcote) [inline requirements.txt]: Okay actually we aren't missing too many deps. I'll put in all the Flask-pyoidc deps for now and shore up the rest later.

**Pull request merged**